x/programs/rust/examples/cpmm/src/lib.rs

```diff
@@ -0,0 +1,304 @@
+use wasmlanche_sdk::{public, state_keys, Context, Program};
+
+#[state_keys]
+pub enum StateKeys {
+    // Internal accounting
+    ReserveX,
+    ReserveY,
+
+    // Liquidity token
+    TotalySupply,
+}
+
+#[public]
+pub fn add_liquidity(context: Context<StateKeys>, amount_x: u64, amount_y: u64) -> u64 {
+    let program = context.program();
+    let total_supply = total_supply(program);
+    // tokens    | shares
+    // -------------------------
+    // amount_x  | minted
+    // reserve_x | total_supply
+    let (reserve_x, reserve_y) = reserves(program);
+    let minted = if total_supply == 0 {
+        let minted = amount_x;
+        assert_eq!(minted, amount_y);
+        minted
+    } else {
+        let minted = amount_x * total_supply / reserve_x;
+        assert_eq!(minted, amount_y * total_supply / reserve_y); // make sure that the ratio is good
+        minted
+    };
+    program
+        .state()
+        .store(StateKeys::ReserveX, &(reserve_x + amount_x))
+        .unwrap();
+    program
+        .state()
+        .store(StateKeys::ReserveY, &(reserve_y + amount_y))
+        .unwrap();
+    program
+        .state()
+        .store(StateKeys::TotalySupply, &(total_supply + minted))
+        .unwrap();
+    minted
+}
+
+#[public]
+pub fn remove_liquidity(context: Context<StateKeys>, shares: u64) -> (u64, u64) {
+    let program = context.program();
+    let total_supply = total_supply(program);
+    let (reserve_x, reserve_y) = reserves(program);
+    let (amount_x, amount_y) = (
+        shares * reserve_x / total_supply,
+        shares * reserve_y / total_supply,
+    );
+    program
+        .state()
+        .store(StateKeys::ReserveX, &(reserve_x - amount_x))
+        .unwrap();
+    program
+        .state()
+        .store(StateKeys::ReserveY, &(reserve_y - amount_y))
+        .unwrap();
+    program
+        .state()
+        .store(StateKeys::TotalySupply, &(total_supply - shares))
+        .unwrap();
+    (amount_x, amount_y)
+}
+
+#[public]
+pub fn swap(context: Context<StateKeys>, amount_in: u64, x_to_y: bool) -> u64 {
+    let program = context.program();
+    let total_supply = total_supply(program);
+    assert!(total_supply > 0, "no liquidity");
+    // k = x * y
+    // x' * y' = x * y
+    // (x + dx) * (y - dy) = x * y
+    // y - dy = (x * y) / (x + dx)
+    // dy = y - (x * y) / (x + dx)
+    // dy = y * dx / (x + dx)
```

Comment:
```suggestion
    // x * y = constant 
    // x' = x + dx
    // y' = y + dy
    // (x + dx) * (y + dy) = x * y
    // y + dy = (x * y) / (x + dx)
    // dy = ((x * y) / (x + dx)) - y
    // skip a few steps 
    // -dy = y * dx / (x + dx)
```
I think it's clearer this way that `dy` and `dx` have opposite signs